File: app/schemas/order.py
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

ALLOWED_CARGO_TYPES = {
    "document",
    "parcel",
    "food",
    "fragile",
    "electronics",
    "clothes",
    "other",
}

ALLOWED_ORDER_STATUSES = {
    "draft",
    "published",
    "bidding",
    "accepted",
    "picked_up",
    "in_transit",
    "delivered",
    "confirmed",
    "cancelled",
    "disputed",
}


class ClientOrderCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    from_city_id: int
    to_city_id: int
    from_district_id: int | None = None
    to_district_id: int | None = None
    pickup_address: str = Field(min_length=1, max_length=1024)
    dropoff_address: str = Field(min_length=1, max_length=1024)
    pickup_lat: Decimal | None = Field(default=None, ge=-90, le=90)
    pickup_lng: Decimal | None = Field(default=None, ge=-180, le=180)
    dropoff_lat: Decimal | None = Field(default=None, ge=-90, le=90)
    dropoff_lng: Decimal | None = Field(default=None, ge=-180, le=180)
    sender_phone: str = Field(min_length=1, max_length=32)
    receiver_phone: str = Field(min_length=1, max_length=32)
    cargo_type: str | None = Field(default=None, max_length=32)
    cargo_photo_url: str | None = Field(default=None, max_length=1024)
    client_price: Decimal | None = Field(default=None, gt=0, le=Decimal("1000000000"))
    comment: str | None = None

    @field_validator("cargo_type")
    @classmethod
    def validate_cargo_type(cls, value: str | None) -> str | None:
        if value is None or value == "":
            return None
        if value not in ALLOWED_CARGO_TYPES:
            raise ValueError("Invalid cargo type")
        return value

    @model_validator(mode="after")
    def validate_coordinate_pairs(self) -> "ClientOrderCreate":
        if (self.pickup_lat is None) != (self.pickup_lng is None):
            raise ValueError("pickup_lat and pickup_lng must be provided together")
        if (self.dropoff_lat is None) != (self.dropoff_lng is None):
            raise ValueError("dropoff_lat and dropoff_lng must be provided together")
        return self


class ClientOrderCancel(BaseModel):
    reason: str = Field(min_length=1)


class SelectDriverRequest(BaseModel):
    bid_id: int


class ClientOrderRatingCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    rating: int
    comment: str | None = None

File: android-app/src/types/order.ts
export type OrderStatus =
  | "draft"
  | "published"
  | "bidding"
  | "accepted"
  | "picked_up"
  | "in_transit"
  | "delivered"
  | "confirmed"
  | "cancelled"
  | "disputed";

export type CreateOrderPayload = {
  from_city_id: number;
  to_city_id: number;
  from_district_id?: number | null;
  to_district_id?: number | null;
  pickup_address: string;
  dropoff_address: string;
  pickup_lat?: number | null;
  pickup_lng?: number | null;
  dropoff_lat?: number | null;
  dropoff_lng?: number | null;
  sender_phone: string;
  receiver_phone: string;
  cargo_type?: string | null;
  cargo_photo_url?: string | null;
  client_price?: number | null;
  comment?: string | null;
};

export type OrderDraft = {
  from_city_id?: string;
  from_city_name?: string;
  from_city_requires_district?: boolean;
  from_district_id?: string | null;
  from_district_name?: string | null;
  to_city_id?: string;
  to_city_name?: string;
  to_city_requires_district?: boolean;
  to_district_id?: string | null;
  to_district_name?: string | null;
  pickup_address?: string;
  dropoff_address?: string;
  pickup_lat?: number | null;
  pickup_lng?: number | null;
  dropoff_lat?: number | null;
  dropoff_lng?: number | null;
  sender_phone?: string;
  receiver_phone?: string;
  cargo_photo_url?: string;
  comment?: string;
};

export type ClientOrder = {
  id: number;
  order_number?: string;
  from_city?: unknown;
  to_city?: unknown;
  from_district?: unknown;
  to_district?: unknown;
  status: OrderStatus;
  suggested_price?: number | null;
  client_price?: number | null;
  final_price?: number | null;
  bids_count?: number;
  cargo_photo_url?: string | null;
  pickup_lat?: number | string | null;
  pickup_lng?: number | string | null;
  dropoff_lat?: number | string | null;
  dropoff_lng?: number | string | null;
  created_at?: string;
};

export type RatingPayload = {
  rating: number;
  comment?: string | null;
};
